### PR TITLE
Move random instant generating code to be a method on interval.Interval.

### PR DESCRIPTION
**Pull request by taknira**

### CODE REVIEW

File: interval/interval.go [hunk 1]
```diff
@@ -15,9 +15,38 @@
...
+// the Interval.
+//
+// If the Interval is nil, or Interval.End <= Interval.Start, the zero value
+// time.Time will be returned.
+func (i *Interval) RandomInstant() time.Time {
+	if i == nil {
+		return time.Time{}
+	}
+
+	if i.Start.Equal(i.End) || i.Start.After(i.End) {
```
Comment (phad) [inline interval/interval.go]: if `i.Start == i.End` then wouldn't returning i.EitherOfThem be good enough? It is as random as you can get within a zero-width interval

File: interval/interval.go [hunk 2]
```diff
@@ -15,9 +15,38 @@
...
 }
+
+// RandomInstant returns a random time (to second precision) that falls within
+// the Interval.
+//
+// If the Interval is nil, or Interval.End <= Interval.Start, the zero value
+// time.Time will be returned.
+func (i *Interval) RandomInstant() time.Time {
+	if i == nil {
+		return time.Time{}
```
Comment (phad) [inline interval/interval.go]: This is nice, and almost lets you omit the `if instant != nil` check in the calling code.
Except, you still want to fall back on the timeNow + validityPeriod default in that case.
I can think of ways around that but they make things less readable, so keep as-is.

File: interval/interval.go [hunk 3]
```diff
@@ -15,9 +15,38 @@
...
+
+	if i.Start.Equal(i.End) || i.Start.After(i.End) {
+		return time.Time{}
+	}
+
+	start := i.Start.Unix()
+	end := i.End.Unix()
+	delta := end - start
+
+	if delta == 0 {
```
Comment (phad) [inline interval/interval.go]: Oh look!  You deal with the `iStart == iEnd` case right here.
How about above, just returning `time{}` for the `iEnd < iStart` edge case.

File: interval/interval_test.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+import (
+	"testing"
+	"time"
+)
+
+func TestRandomInstant(t *testing.T) {
+	tests := []struct {
+		desc     string
+		iv       *Interval
```
Comment (phad) [inline interval/interval_test.go]: eek! initialization vector.  Not sure what else you can call this.  Sometimes I use 'in' indicating 'input'.

File: interval/interval.go (hunk 1 above)
Comment (taknira) [reply interval/interval.go]: No - if you look at the documentation for the Interval type, the interval is [Start, End), so if Start == End there is nothing in the interval/the interval is invalid (how do you include Start but exclude End when they are the same?).

File: interval/interval.go (hunk 3 above)
Comment (taknira) [reply interval/interval.go]: This is a nuance, and maybe it's not quite right/is complicating things unecessarily - if we've got here then Interval.Start != Interval.End, but Interval.Start.Unix() == Interval.End.Unix(), implying there is less than a second's difference between Interval.Start and Interval.End, as the .Unix() is only to second precision.

Comment (taknira) [reply interval/interval.go]: Might have another think about this change, and even if it stays this way I'll add some comments.

Comment (taknira) [reply interval/interval.go]: Yeah, there's something here that's incorrect (have found a test case that fails) - will fix things and let you know when it's better :)

File: interval/interval.go (hunk 1 above)
Comment (phad) [reply interval/interval.go]: I overlooked that it is [start, end) - ok then.

File: interval/interval_test.go (hunk 4 above)
Comment (taknira) [reply interval/interval_test.go]: Changed :)

File: interval/interval.go (hunk 3 above)
Comment (taknira) [reply interval/interval.go]: Okay I think I've fixed it.  I've also added comments explaining what's going on, and I've renamed the method to RandomSecond as it chooses a specific second.  The reason I need it to do that is because the NotAfter field in certificates (which this is being used for) are specified to second precision.

File: interval/interval.go [hunk 5]
```diff
@@ -43,8 +55,15 @@ func (i *Interval) RandomInstant() time.Time {
 	end := i.End.Unix()
```
Comment (phad) [inline interval/interval.go]: (this is about line 50 but I can't attach a comment there ?!)

`if i.Start.Equal(i.End) || i.Start.After(i.End)` is equivalent to `if !i.Start.Before(i.End)`

File: interval/interval.go [hunk 6]
```diff
@@ -43,8 +55,15 @@ func (i *Interval) RandomInstant() time.Time {
...
 	delta := end - start
 
+	// If delta == 0 there is less than 1 second between i.Start and i.End.
 	if delta == 0 {
-		return i.Start
+		// If there is a second-precision time between i.Start and i.End,
+		// return that.
+		if t := time.Unix(end, 0); !t.Before(i.Start) {
```
Comment (phad) [inline interval/interval.go]: Maybe it's Friday and I'm being slow, but... can you give an example for this. (perhaps you have a test case).

Comment (phad) [reply interval/interval.go]: so, given that `t` is `end` truncated to second-precision;  and `t` is either equal to `i.Start` or after `i.Start` (which is not truncated so can have a nanos component), then you deem `t` within the interval and return it.

Comment (phad) [reply interval/interval.go]: And this is the case where `i.End-i.Start` is both sub-second and within the same second.

Comment (phad) [reply interval/interval.go]: If I got that right, the final statement above seems like a decent comment for it.

File: interval/interval.go (hunk 5 above)
Comment (taknira) [reply interval/interval.go]: Ahaha yup.  Done.

File: interval/interval.go (hunk 6 above)
Comment (taknira) [reply interval/interval.go]: The logic was wrong again :(  Third time's a charm?
